test/integration/scheduler/taint_test.go

```diff
@@ -198,116 +250,142 @@ func TestTaintNodeByCondition(t *testing.T) {
+					Effect: v1.TaintEffectNoSchedule,
 				},
 			},
 		},
 	}
 
-	if _, err := clientset.CoreV1().Pods(nsName).Create(networkDaemonPod); err != nil {
-		t.Errorf("Case 4: Failed to create pod for network daemon: %v", err)
-	} else {
-		if err := waitForPodToScheduleWithTimeout(clientset, networkDaemonPod, time.Second*60); err != nil {
-			t.Errorf("Case 4: Failed to schedule network daemon pod in 60s.")
-		}
-	}
-
-	// Case 5: Taint node by unschedulable condition
-	unschedulableNode := &v1.Node{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "node-2",
-		},
-		Spec: v1.NodeSpec{
-			Unschedulable: true,
-		},
-		Status: v1.NodeStatus{
-			Capacity: v1.ResourceList{
-				v1.ResourceCPU:    resource.MustParse("4000m"),
-				v1.ResourceMemory: resource.MustParse("16Gi"),
-				v1.ResourcePods:   resource.MustParse("110"),
+	for i, test := range tests {
+		node := &v1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node-1",
 			},
-			Allocatable: v1.ResourceList{
-				v1.ResourceCPU:    resource.MustParse("4000m"),
-				v1.ResourceMemory: resource.MustParse("16Gi"),
-				v1.ResourcePods:   resource.MustParse("110"),
+			Spec: v1.NodeSpec{
+				Unschedulable: test.unschedulable,
 			},
-		},
-	}
+			Status: v1.NodeStatus{
+				Capacity:    nodeRes,
+				Allocatable: nodeRes,
+				Conditions:  test.nodeConditions,
+			},
+		}
+
+		if _, err := cs.CoreV1().Nodes().Create(node); err != nil {
+			t.Errorf("Case %d (%s): Failed to create node, err: %v", i, test.name, err)
+		}
+
+		if err := waitForNodeTaints(cs, node, test.expectedTaints); err != nil {
+			t.Errorf("Case %d (%s): Failed to taint node, err: %v", i, test.name, err)
+		}
+
+		var pods []*v1.Pod
 
-	nodeInformerCh2 := make(chan bool)
-	nodeInformer2 := informers.Core().V1().Nodes().Informer()
-	nodeInformer2.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		UpdateFunc: func(old, cur interface{}) {
-			curNode := cur.(*v1.Node)
-			if curNode.Name != "node-2" {
-				return
+		for j, p := range test.pods {
+			pod := p.pod.DeepCopy()
+			pod.Name = fmt.Sprintf("%s-%d", pod.Name, j)
+			pod.Spec.Tolerations = p.tolerations
+
+			createdPod, err := cs.CoreV1().Pods(pod.Namespace).Create(pod)
+			if err != nil {
+				t.Fatalf("Case %d (%s): Failed to create pod %s/%s, error: %v",
+					i, test.name, pod.Namespace, pod.Name, err)
 			}
 
-			for _, taint := range curNode.Spec.Taints {
-				if taint.Key == algorithm.TaintNodeUnschedulable &&
-					taint.Effect == v1.TaintEffectNoSchedule {
-					nodeInformerCh2 <- true
-					break
+			pods = append(pods, createdPod)
+
+			if p.fits {
+				if err := waitForPodToSchedule(cs, createdPod); err != nil {
+					t.Errorf("Case %d (%s): Failed to schedule pod %s/%s on the node, err: %v", i, test.name,
+						pod.Namespace, pod.Name, err)
+				}
+			} else {
+				if err := waitForPodUnschedulable(cs, createdPod); err != nil {
+					t.Errorf("Case %d (%s): Failed to unschedule pod %s/%s on the node, err: %v", i, test.name,
```

Comment:
nit: the msg seems confusing to me, how about something like "unscheduble pod <pod-name> gets scheduled"?

Reply:
+1 to aveshagarwal's comment